boot/boot_test.go

```diff
@@ -3749,3 +3764,138 @@ func (s *bootKernelCommandLineSuite) TestCommandLineUpdateUC20TransitionFullExtr
+		"snapd_recovery_mode=run static mocked panic=-1",
+		"snapd_recovery_mode=run static mocked panic=-1 extra args",
+	}})
+	// bootenv hasn't been updated yet
+	c.Check(s.bootloader.SetBootVarsCalls, Equals, 0)
+	// but modeenv has already been updated
+	m, err := boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	c.Check(m.CurrentKernelCommandLines, DeepEquals, boot.BootCommandLines{
+		"snapd_recovery_mode=run static mocked panic=-1",
+		"snapd_recovery_mode=run static mocked panic=-1 extra args",
+	})
+
+	// REBOOT
+	resealPanic = false
+	err = boot.MarkBootSuccessful(s.uc20dev)
+	c.Assert(err, IsNil)
+	// we resealed after reboot, since modeenv was updated and carries the
+	// current command line only
+	c.Check(s.resealCalls, Equals, 2)
+	m, err = boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	c.Check(m.CurrentKernelCommandLines, DeepEquals, boot.BootCommandLines{
+		"snapd_recovery_mode=run static mocked panic=-1",
+	})
+
+	// try the update again, but no panic in reseal this time
+	s.resealCalls = 0
+	s.resealCommandLines = nil
+	resealPanic = false
+	// but panic in set
+	c.Assert(func() {
+		boot.UpdateCommandLineForGadgetComponent(s.uc20dev, sf)
+	}, PanicMatches, "mocked reboot panic in SetBootVars")
+	c.Check(s.resealCalls, Equals, 1)
+	c.Check(s.resealCommandLines, DeepEquals, [][]string{{
+		// those come from boot chains which use predictable sorting
+		"snapd_recovery_mode=run static mocked panic=-1",
+		"snapd_recovery_mode=run static mocked panic=-1 extra args",
+	}})
+	// the call to bootloader wasn't counted, because it called panic
+	c.Check(s.bootloader.SetBootVarsCalls, Equals, 0)
+	m, err = boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	c.Check(m.CurrentKernelCommandLines, DeepEquals, boot.BootCommandLines{
+		"snapd_recovery_mode=run static mocked panic=-1",
+		"snapd_recovery_mode=run static mocked panic=-1 extra args",
+	})
+
+	// REBOOT
+	err = boot.MarkBootSuccessful(s.uc20dev)
+	c.Assert(err, IsNil)
+	// we resealed after reboot again
+	c.Check(s.resealCalls, Equals, 2)
+	m, err = boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	c.Check(m.CurrentKernelCommandLines, DeepEquals, boot.BootCommandLines{
+		"snapd_recovery_mode=run static mocked panic=-1",
+	})
+
+	// try again, for the last time, things should go smoothly
+	s.resealCalls = 0
+	s.resealCommandLines = nil
+	restoreBootloaderNoPanic()
+	reboot, err := boot.UpdateCommandLineForGadgetComponent(s.uc20dev, sf)
+	c.Assert(err, IsNil)
+	c.Check(reboot, Equals, true)
+	c.Check(s.resealCalls, Equals, 1)
+	c.Check(s.bootloader.SetBootVarsCalls, Equals, 1)
+	// all done, modeenv
+	m, err = boot.ReadModeenv("")
+	c.Assert(err, IsNil)
+	c.Check(m.CurrentKernelCommandLines, DeepEquals, boot.BootCommandLines{
+		"snapd_recovery_mode=run static mocked panic=-1",
+		"snapd_recovery_mode=run static mocked panic=-1 extra args",
+	})
+	args, err := s.bootloader.GetBootVars("snapd_extra_cmdline_args", "snapd_full_cmdline_args")
```

Comment:
Should we check earlier that the canary is still intact in this test?

Reply:
Earlier bits either panic'ed before calling SetBootVars or right when entering it (due to SetMockPanic), in either case the test checks that SetBootVarsCalls is 0.